nixos/ollama: add option for `HSA_OVERRIDE_GFX_VERSION`

abysssol opened the pull request:
## Description of changes

There are several GPUs that ROCm doesn't officially support but will work correctly if ROCm is directed to treat the GPU as a different one that is supported and has a similar architecture.

This can be done by setting `HSA_OVERRIDE_GFX_VERSION`. Ollama has documentation on this topic: https://github.com/ollama/ollama/blob/main/docs/gpu.md#amd-radeon

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [x] `nix build -f pkgs/top-level/release.nix release-checks`
  - [x] `nix build -f default.nix ollama.tests`
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix run nixpkgs#nixpkgs-review -- rev HEAD`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Aleksanaa commented on pkgs/by-name/ol/ollama/package.nix:
```diff
@@ -185,7 +185,8 @@ goBuild ((lib.optionalAttrs enableRocm {
       lib.optionalString enableRocm
-        ''--set-default HIP_PATH '${rocmPath}' ''} \
+        ''--set-default HIP_PATH '${rocmPath}' \
+          --set-default HSA_OVERRIDE_GFX_VERSION 0 ''} \
```
What does 0 mean here?

drupol replied:
Ho good catch !

abysssol replied:
It's essentially garbage. It could just as well be a "1" or "." or "-". I added it because of its interaction with rocblas compiled without tensile files.

See #321335 and #320202 for details, tldr rocblas compiled with tensile files (large lazy loaded gpu specific data (and/or libraries?)) takes a lot of processing, memory, and storage to compile, and hydra won't cache it because it's too large or something like that. Most ollama users are unable to compile it, their machines choking and even crashing. So, rocblas was recently changed back to not compile tensile files, which works fine for some gpus, but for others (including mine), rocm becomes unable to detect the gpu unless HSA_OVERRIDE_GFX_VERSION is set.

I very recently figured out that the actual value of HSA_OVERRIDE_GFX_VERSION doesn't actually matter for my gpu (6950 xt), and any nonempty string will enable rocm for gfx 1030, while an empty string or being unset will cause ollama to only use cpu. Thus the garbage value of "0" causes rocm to work correctly for me without any further configuration, and my hope is that this will also apply to other users, allowing them to use rocm without having to figure out some arcane override (it's not like I really care about my own setup working with ollama, I don't actually even use ollama).

Aleksanaa replied:
> rocm becomes unable to detect the gpu unless HSA_OVERRIDE_GFX_VERSION is set

Could you add this to a comment near it?

abysssol commented:
@kira-bruneau since your gpu doesn't work with rocm without HSA_OVERRIDE_GFX_VERSION, could you test ollama-rocm from this pr and see if rocm works correctly without you setting HSA_OVERRIDE_GFX_VERSION manually?

See my [earlier comment](https://github.com/NixOS/nixpkgs/pull/321664#discussion_r1649727925) for details, but essentially HSA_OVERRIDE_GFX_VERSION needs to be unset (**not** just set to an empty string) for the default value to be set in ollama's wrapper script. In bash thats `unset HSA_OVERRIDE_GFX_VERSION`, in fish it's `set -e HSA_OVERRIDE_GFX_VERSION`.

Alternatively, you could set HSA_OVERRIDE_GFX_VERSION to "0", the same as what the wrapper would use as a default, if that's more convenient.

You may also need to set `OLLAMA_HOST=127.0.0.1:11435` if you already have the ollama service running.

I'm hoping since both our gpus aren't used by rocm without setting HSA_OVERRIDE_GFX_VERSION, and mine works even when setting it to a garbage value, your gpu will also work without needing to manually define HSA_OVERRIDE_GFX_VERSION as a valid gfx version. If so, then that's good evidence that many other users' gpus will behave similarly, making using rocm easier for many more people.

kira-bruneau commented:
Oh I just tried and ended up with a coredump:

```
Error: llama runner process has terminated: signal: aborted (core dumped) error:Could not initialize Tensile host: No devices found
```

I think it would be better to leave it up to the user to configure this override.

abysssol commented:
@kira-bruneau I see ... that's unfortunate. Thank you for testing it!

abysssol committed:
nixos/ollama: add option for `HSA_OVERRIDE_GFX_VERSION`

There are several GPUs that ROCm doesn't officially support but
will work correctly if ROCm is directed to treat the GPU as a different
one that is supported and has a similar architecture.
This can be done by setting `HSA_OVERRIDE_GFX_VERSION`.
Ollama has documentation on this topic: https://github.com/ollama/ollama/blob/main/docs/gpu.md#amd-radeon

abysssol committed:
ollama: refactor wrapper options

drupol reviewed (approved):
LGTM

kira-bruneau commented:
Result of `nixpkgs-review pr 321664` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>ollama-cuda</li>
    <li>ollama-rocm</li>
  </ul>
</details>

caffineehacker commented:
FYI - I think this may have broken existing users that set environmentVariables.HSA_OVERRIDE_GFX_VERSION since this now overwrites that value rather than defaulting to what the user had already set.

I believe using mkMerge will work correctly, but I'll need to do some testing. I'll try to get a PR tonight if I have time, but anyone else can feel free to fix this sooner.

Aleksanaa commented:
> FYI - I think this may have broken existing users that set environmentVariables.HSA_OVERRIDE_GFX_VERSION since this now overwrites that value rather than defaulting to what the user had already set.

It only sets the value when the new option is set to something other than `null`.

caffineehacker commented:
> > FYI - I think this may have broken existing users that set environmentVariables.HSA_OVERRIDE_GFX_VERSION since this now overwrites that value rather than defaulting to what the user had already set.
> 
> It only sets the value when the new option is set to something other than `null`.

Unfortunately that's not how lib.mkIf works. I believe you need to use lib.mkMerge if you're using lib.mkIf. For example, try running this in nix repl:

```
nix-repl> ({ b = 1; } // { b = lib.mkIf false 2; }).b
{ _type = "if"; condition = false; content = 2; }
```

b is no longer 1 even though lib.mkIf had false as the first argument. I won't have time to send a PR tonight, but I should tomorrow.
